### PR TITLE
[SPARK-33513][BUILD] Upgrade to Scala 2.13.4 to improve exhaustivity

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims the followings.
1. Upgrade from Scala 2.13.3 to 2.13.4 for Apache Spark 3.1
2. Fix exhaustivity issues in both Scala 2.12/2.13 (Scala 2.13.4 requires this for compilation.)
3. Enforce the improved exhaustive check by using the existing Scala 2.13 GitHub Action compilation job.

### Why are the changes needed?

Scala 2.13.4 is a maintenance release for 2.13 line and improves JDK 15 support.
- https://github.com/scala/scala/releases/tag/v2.13.4

Also, it improves exhaustivity check.
- https://github.com/scala/scala/pull/9140 (Check exhaustivity of pattern matches with "if" guards and custom extractors)
- https://github.com/scala/scala/pull/9147 (Check all bindings exhaustively, e.g. tuples components)

### Does this PR introduce _any_ user-facing change?

Yep. Although it's a maintenance version change, it's a Scala version change.

### How was this patch tested?

Pass the CIs and do the manual testing.
- Scala 2.12 CI jobs(GitHub Action/Jenkins UT/Jenkins K8s IT) to check the validity of code change.
- Scala 2.13 Compilation job to check the compilation

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
This PR is ready for review finally.
Could you review this PR, @srowen , @HyukjinKwon , @maropu , @LuciferYang ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 1]
```diff
@@ -322,7 +322,9 @@ case class Literal (value: Any, dataType: DataType) extends LeafExpression {
         case (a: Array[Byte], b: Array[Byte]) => util.Arrays.equals(a, b)
         case (a: ArrayBasedMapData, b: ArrayBasedMapData) =>
           a.keyArray == b.keyArray && a.valueArray == b.valueArray
-        case (a, b) => a != null && a.equals(b)
+        case (a: Double, b: Double) if a.isNaN && b.isNaN => true
+        case (a: Float, b: Float) if a.isNaN && b.isNaN => true
+        case (a, b) => a != null && a == b
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:327]: (Just a question) Are the changes above related to this PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -362,7 +362,7 @@ case class Join(
         left.constraints
       case RightOuter =>
         right.constraints
-      case FullOuter =>
+      case _ =>
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:365]: nit: for readability, how about leaving the comment like this?
```
case _ => // FullOuter
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala [hunk 3]
```diff
@@ -120,7 +120,7 @@ class GenericArrayData(val array: Array[Any]) extends ArrayData {
             if (!o2.isInstanceOf[Double] || ! java.lang.Double.isNaN(o2.asInstanceOf[Double])) {
               return false
             }
-          case _ => if (!o1.equals(o2)) {
+          case _ => if (o1.getClass != o2.getClass || o1 != o2) {
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala:123]: ditto: Is this change above related to this PR?

File: streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala [hunk 4]
```diff
@@ -293,7 +293,7 @@ private[streaming] object FileBasedWriteAheadLog {
           val startTime = startTimeStr.toLong
           val stopTime = stopTimeStr.toLong
           Some(LogInfo(startTime, stopTime, file.toString))
-        case None =>
+        case None | Some(_) =>
```
Comment (maropu) [inline streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala:296]: We cannot write it like `case _ =>` here?

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala [hunk 5]
```diff
@@ -757,15 +757,15 @@ private[spark] object JsonProtocol {
 
   def taskResourceRequestMapFromJson(json: JValue): Map[String, TaskResourceRequest] = {
     val jsonFields = json.asInstanceOf[JObject].obj
-    jsonFields.map { case JField(k, v) =>
+    jsonFields.collect { case JField(k, v) =>
```
Comment (maropu) [inline core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:760]: Ah, I see.  The compilation of `map` in this case fails in v2.13.4?

**Review by maropu — APPROVED** [review]
I left minor comments (& question) and the other parts look fine.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:327]: Yes. It's a compilation error. We cannot use `.equals` for `Any` type.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:365]: Ya. I thought like that, but actually, there are many missing patterns.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:365]: It was difficult to track one-by-one. Logically, this is the rest of the previous patterns. So, I decided to skip that comment.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala:123]: Yes. We cannot use `equals` and we cannot use `!=` here due to the following.
```scala
scala> Float.NaN == Float.NaN
val res2: Boolean = false

scala> Float.NaN.equals(Float.NaN)
val res3: Boolean = true
```

File: core/src/main/scala/org/apache/spark/util/JsonProtocol.scala (hunk 5 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/JsonProtocol.scala:760]: Yes.

File: streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply streaming/src/main/scala/org/apache/spark/streaming/util/FileBasedWriteAheadLog.scala:296]: We can use like that~

**Comment by dongjoon-hyun** [comment]
The new pattern match is strict and the `master` branch change can easily break it. I'll merge this first with @maropu 's approval and do the follow-up.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala (hunk 1 above)
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala:327]: Oh, I see.

**Comment by dongjoon-hyun** [comment]
Thank you always, @maropu !

**Review by srowen — COMMENTED** [review]
Late LGTM

**Review by HyukjinKwon — COMMENTED** [review]
LGTM

**Comment by dongjoon-hyun** [comment]
Thank you, @srowen and @HyukjinKwon .

File: mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala [hunk 6]
```diff
@@ -298,7 +299,10 @@ private[ml] object RFormulaParser extends RegexParsers {
   private val expr = (sum | term)
 
   private val formula: Parser[ParsedRFormula] =
-    (label ~ "~" ~ expr) ^^ { case r ~ "~" ~ t => ParsedRFormula(r, t.asTerms.terms) }
+    (label ~ "~" ~ expr) ^^ {
+      case r ~ "~" ~ t => ParsedRFormula(r, t.asTerms.terms)
```
Comment (retronym) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala:303]: For future reference, a slightly cleaner way to express this in a warning free way is:

```
((label <~ "~") ~ expr ) ^^ {
  case r ~ t => ParsedRFormula(r, t.asTerms.terms)
}
```

It also avoids duplicating the `"~"` token in the parser and the action.

Comment (dongjoon-hyun) [reply mllib/src/main/scala/org/apache/spark/ml/feature/RFormulaParser.scala:303]: Feel free to make a PR, @retronym . :)
